### PR TITLE
refactor index validation and prevent indexes over json columns

### PR DESCRIPTION
**Pull request by jycor**
This PR consolidates the logic to validate if an index.
Additionally, it fixes a bug where `create table t (i int, index (i, i));` was allowed.

fixes: https://github.com/dolthub/dolt/issues/6064

### CODE REVIEW

File: sql/rowexec/ddl_iters.go [hunk 1]
```diff
@@ -1806,74 +1806,79 @@ func getCheckAlterableTable(t sql.Table) (sql.CheckAlterableTable, error) {
 	}
 }
 
+// generateIndexName generates a unique index name based on the columns in the index, and any existing indexes on the table.
+func generateIndexName(ctx *sql.Context, idxAltable sql.IndexAlterableTable, idxColNames []string) (string, error) {
```
Comment (max-hoffman) [inline sql/rowexec/ddl_iters.go:1810]: the comment above that this is different than Dolt's implementation is maybe informative to keep?

File: sql/analyzer/validate_create_table.go [hunk 2]
```diff
@@ -836,6 +778,7 @@ func removeInSchema(sch sql.Schema, colName, tableName string) sql.Schema {
 }
 
 // TODO: make this work for CREATE TABLE statements where there's a non-pk auto increment column
+// TODO: verify that the above is fixed
```
Comment (max-hoffman) [inline sql/analyzer/validate_create_table.go:781]: if it's fresh in your head and easy we could just add it

File: enginetest/enginetests.go [hunk 3]
```diff
@@ -5559,9 +5559,14 @@ func findRole(toUser string, roles []*mysql_db.RoleEdge) *mysql_db.RoleEdge {
 func TestBlobs(t *testing.T, h Harness) {
 	h.Setup(setup.MydbData, setup.BlobData, setup.MytableData)
 
+	// By default, strict_mysql_compatibility is disabled, but these tests require it to be enabled.
+	err := sql.SystemVariables.SetGlobal("strict_mysql_compatibility", int8(1))
+	require.NoError(t, err)
 	for _, tt := range queries.BlobErrors {
 		runQueryErrorTest(t, h, tt)
 	}
+	err = sql.SystemVariables.SetGlobal("strict_mysql_compatibility", int8(0))
```
Comment (max-hoffman) [inline enginetest/enginetests.go:5568]: setting in a defer might be a bit safer

**Review by max-hoffman — APPROVED** [review]
lgtm, all seems reasonable!

File: sql/analyzer/validate_create_table.go (hunk 2 above)
Comment (jycor) [reply sql/analyzer/validate_create_table.go]: double checked, and there are existing tests for this case that pass
https://github.com/dolthub/go-mysql-server/blob/main/enginetest/queries/create_table_queries.go#L120

File: enginetest/enginetests.go (hunk 3 above)
Comment (jycor) [reply enginetest/enginetests.go:5568]: I only wanted `queries.BlobErrors` to do this; I think some of the tests that come later might want the variable to be set to false.

File: sql/rowexec/ddl_iters.go (hunk 1 above)
Comment (jycor) [reply sql/rowexec/ddl_iters.go:1810]: I can't find the deleted comment, but I think I might've been mistaken before.

**Pull request merged**